spec/unit/pops/types/p_object_type_spec.rb

```diff
@@ -0,0 +1,599 @@
+        attributes => {
+          a => { type => Integer[0,10], override => true }
+        }
+      OBJECT
+      expect_object('MyObject', parent)
+      tp = parse_object('MyDerivedObject', obj)
+      expect(tp['a'].type).to eql(PIntegerType.new(0,10))
+    end
+
+    it 'raises an error when the an attribute overrides a function' do
+      parent = <<-OBJECT
+        attributes => {
+          a => Integer
+        }
+      OBJECT
+      obj = <<-OBJECT
+        parent => MyObject,
+        functions => {
+          a => { type => Callable, override => true }
+        }
+      OBJECT
+      expect_object('MyObject', parent)
+      expect { parse_object('MyDerivedObject', obj) }.to raise_error(Puppet::ParseError,
+        "function 'a' overrides inherited attribute")
+    end
+
+    it 'raises an error when the an function overrides an attribute' do
+      parent = <<-OBJECT
+        functions => {
+          a => Callable
+        }
+      OBJECT
+      obj = <<-OBJECT
+        parent => MyObject,
+        attributes => {
+          a => { type => Integer, override => true }
+        }
+      OBJECT
+      expect_object('MyObject', parent)
+      expect { parse_object('MyDerivedObject', obj) }.to raise_error(Puppet::ParseError,
+        "attribute 'a' overrides inherited function")
+    end
+
+    it 'raises an error on attempts to redefine inherited member to unassignable type' do
+      parent = <<-OBJECT
+        attributes => {
+          a => Integer
+        }
+      OBJECT
+      obj = <<-OBJECT
+        parent => MyObject,
+        attributes => {
+          a => { type => String, override => true }
+        }
+      OBJECT
+      expect_object('MyObject', parent)
+      expect { parse_object('MyDerivedObject', obj) }.to raise_error(Puppet::ParseError,
+        "attempt to override attribute 'a' with a type that does not match")
+    end
+
+    it 'raises an error when the an attribute overrides a final attribute' do
+      parent = <<-OBJECT
+        attributes => {
+          a => { type => Integer, final => true }
+        }
+      OBJECT
+      obj = <<-OBJECT
+        parent => MyObject,
+        attributes => {
+          a => { type => Integer, override => true }
+        }
+      OBJECT
+      expect_object('MyObject', parent)
+      expect { parse_object('MyDerivedObject', obj) }.to raise_error(Puppet::ParseError,
+        "attempt to override final attribute 'a'")
+    end
+
+    it 'raises an error when the an overriding attribute is not declared with override => true' do
+      parent = <<-OBJECT
+        attributes => {
+          a => Integer
+        }
+      OBJECT
+      obj = <<-OBJECT
+        parent => MyObject,
+        attributes => {
+          a => Integer
+        }
+      OBJECT
+      expect_object('MyObject', parent)
+      expect { parse_object('MyDerivedObject', obj) }.to raise_error(Puppet::ParseError,
+        "attribute 'a' attempts to override without having override defined")
```

Comment:
actually, override must be true - message should say that, override => false is also a definition